### ISSUE TITLE
Adding missing flag icon for language variant

### ISSUE THREAD

**Issue by bad63r** [zen-tools/gxkb#33]
Hello,
I'm trying to add to ~/.local/share/gxkb/flags a flag which is missing for my language. Ruthenian is different language than serbian, but technically in gxkb is done that it is implemented as part of serbian keyboard variant.

` gkbd-keyboard-display -l "rs(rue)"`
This command will give me keyboard layout for my language. No idea though how to add missing flag in gxkb system tray indicator applet.

I tried adding rs(rue).png, rs-rue.png to the  ~/.local/share/gxkb/flags  but no success.

 Any help is appreciated. 

Thanks, Tom

**Comment by zen-tools** [comment]
Hi, Tom!

I just checked `strace -e trace=file gxkb` output, and it says that gxkb is looking for icon under `~/.local/share/gxkb/flags/rs.png` path. While that checking I have used the next gxkb.cfg settings:
```
layouts=us,rs
variants=,rue
```
Also, gxkb may cache default zz.png icon, so you probably may have to restart it.

**Comment by bad63r** [comment]
Oh my bad. I forgot to mention the gist of my problem. I also use rs language without variant, and rs(rue) as its variant. I'm using us,rs,rs(rue) with config that resemble at config you depicted. My goal is somehow to distinguish rs from rs(rue) with flags in system tray icons. 
Once again, sorry that I didn't mention that in the beginning :( 

Tom

**Comment by zen-tools** [comment]
That requires some code changes. I will look how to fix that..

**Comment by pfactum** [comment]
@zen-tools would it be possible to change the way how flag icons are looked up maybe? For instance, if there's `layouts=us,ua,ua` and `variants=altgr-intl,unicode,homophonic`, `gxkb` would look for `ua-unicode.png` and `ua-homophonic.png` files first, and then bail out to using `ua.png` if none are found.

**Comment by zen-tools** [comment]
Hi! Yes, it's possible. Just need to find some time to implement that.

**Comment by pfactum** [comment]
@zen-tools does this work for you? I can make a pull request.

```c
commit b5cfd9e6e7d45e9657ae070277bb9bcc843adbeb
Author: Oleksandr Natalenko <oleksandr@natalenko.name>
Date:   Sun Apr 18 17:42:26 2021 +0200

    support flags for variants
    
    Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>

diff --git a/src/statusicon.c b/src/statusicon.c
index 0acbd4c..20c472a 100644
--- a/src/statusicon.c
+++ b/src/statusicon.c
@@ -128,7 +128,8 @@ void
 statusicon_update_current_image( void )
 {
     const gchar *group_name = xkb_config_get_group_name( -1 );
-    gchar *filepath = xkb_util_get_flag_filename( group_name );
+    const gchar *variant = xkb_config_get_variant( -1 );
+    gchar *filepath = xkb_util_get_flag_filename( group_name, variant );
 
     if( icon_type == SYSTRAY )
     {
@@ -176,7 +177,8 @@ statusicon_update_menu( void )
     {
         gchar *layout_string = NULL;
 
-        gchar *imgfilename = xkb_util_get_flag_filename( xkb_config_get_group_name( i ) );
+        gchar *imgfilename = xkb_util_get_flag_filename( xkb_config_get_group_name( i ),
+                                                         xkb_config_get_variant( i ) );
         GdkPixbuf *handle = gdk_pixbuf_new_from_file_at_scale( imgfilename, width, height, TRUE, NULL );
         g_free( imgfilename );
 
diff --git a/src/xkb-util.c b/src/xkb-util.c
index c08a814..c1669d4 100644
--- a/src/xkb-util.c
+++ b/src/xkb-util.c
@@ -24,39 +24,86 @@
 #include "xkb-util.h"
 
 gchar*
-xkb_util_get_flag_filename( const gchar* group_name )
+xkb_util_get_flag_filename( const gchar* group_name, const gchar* variant )
 {
     gchar* filepath;
 
     if( group_name == NULL )
         return NULL;
 
-    gchar *flag_name = g_ascii_strdown( group_name, -1 );
-    flag_name = g_strstrip( flag_name );
+    gchar* single = g_ascii_strdown( group_name, -1 );
+    single = g_strstrip( single );
 
-    gchar* filename = g_strconcat( flag_name, ".png", NULL );
-    g_free( flag_name );
+    gchar* pair = NULL;
+    if( variant )
+    {
+        gchar* joined = g_strjoin( "-", group_name, variant, NULL );
+        pair = g_ascii_strdown( joined, -1 );
+        g_free( joined );
+        pair = g_strstrip( pair );
+    }
+
+    gchar* filename_single = g_strconcat( single, ".png", NULL );
+    g_free( single );
+
+    gchar* filename_pair = NULL;
+    if( pair )
+    {
+        filename_pair = g_strconcat( pair, ".png", NULL );
+        g_free( pair );
+    }
 
     // Try to get image from user data directory
+    if( filename_pair )
+    {
+        filepath = g_strjoin( "/", xkb_util_get_data_dir(),
+                              "flags", filename_pair, NULL );
+
+        if( g_file_test( filepath, G_FILE_TEST_EXISTS ) )
+        {
+            g_free( filename_pair );
+            g_free( filename_single );
+            return filepath;
+        } else
+            g_free( filepath );
+    }
+
     filepath = g_strjoin( "/", xkb_util_get_data_dir(),
-                          "flags", filename, NULL );
+                          "flags", filename_single, NULL );
 
     if( g_file_test( filepath, G_FILE_TEST_EXISTS ) )
     {
-        g_free( filename );
+        g_free( filename_pair );
+        g_free( filename_single );
         return filepath;
     }
 
     // Try to get image from system directory
-    filepath = g_strjoin( "/", FLAGSDIR, filename, NULL );
+    if( filename_pair )
+    {
+        filepath = g_strjoin( "/", FLAGSDIR, filename_pair, NULL );
+        g_free( filename_pair );
+
+        if( g_file_test( filepath, G_FILE_TEST_EXISTS ) )
+        {
+            g_free( filename_single );
+            return filepath;
+        } else
+            g_free( filepath );
+    }
+
+    filepath = g_strjoin( "/", FLAGSDIR, filename_single, NULL );
 
-    g_free( filename );
+    g_free( filename_single );
 
     if( !g_file_test( filepath, G_FILE_TEST_EXISTS ) )
     {
         filepath = g_strjoin( "/", FLAGSDIR, "zz.png", NULL );
         if( !g_file_test( filepath, G_FILE_TEST_EXISTS ) )
+        {
+            g_free( filepath );
             return NULL;
+        }
     }
 
     return filepath;
diff --git a/src/xkb-util.h b/src/xkb-util.h
index d56e4ba..52767b0 100644
--- a/src/xkb-util.h
+++ b/src/xkb-util.h
@@ -28,7 +28,8 @@
 
 #include <glib.h>
 
-gchar*      xkb_util_get_flag_filename      (const gchar* group_name);
+gchar*      xkb_util_get_flag_filename      (const gchar* group_name,
+                                             const gchar* variant);
 
 gchar*      xkb_util_get_layout_string      (const gchar* group_name,
                                              const gchar* variant);
```

**Comment by zen-tools** [comment]
@pfactum thanks for provided code snipped, I really appreciate this. But I already did the same in the commit d70f5d4, except I have changed delimiter to underscore and add GSList into xkb_util_get_flag_filename function to make code more compact.

**Comment by pfactum** [comment]
@zen-tools your change seems to work for me as well, thanks!
@bad63r hopefully, we are done with it ☺.

**Comment by pfactum** [comment]
@zen-tools btw, shall you tag a new point release so that it would be easier to get this change into a distribution (Arch in my case)? thanks.

**Comment by pfactum** [comment]
@zen-tools also btw, your change doesn't load flags from `FLAGSDIR` any more.

**Comment by zen-tools** [comment]
Nice catch, thanks! It should be fixed now.

Regarding the new release, I will do it later. Let's first make sure that there will be no other issues. :smile:

**Comment by pfactum** [comment]
Seems to work fine with your latest fixes:

```
access("/home/pf/.local/share/gxkb", F_OK) = 0
access("/home/pf/.local/share/gxkb/flags/ua_ula.png", F_OK) = -1 ENOENT (No such file or directory)
access("/usr/share/gxkb/flags/ua_ula.png", F_OK) = 0

access("/home/pf/.local/share/gxkb", F_OK) = 0
access("/home/pf/.local/share/gxkb/flags/ua_uca.png", F_OK) = -1 ENOENT (No such file or directory)
access("/usr/share/gxkb/flags/ua_uca.png", F_OK) = 0
```

Thanks.

**Comment by zen-tools** [comment]
@pfactum I just released v0.9.1, so you can create build for Arch. Also, I am closing this issue, but feel free to reopen it if you find a bug. Have a nice weekend. :slightly_smiling_face:
